sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,318 @@
+/*!
+# Introduction
+
+shibaken is called by sundog as a setting generator.
+Its sole parameter is the user-data provided by the api server.
+
+In the event that user-data is empty and has not been provided by the user, shibaken will generate
+its own user-data by fetching authorized ssh keys from the AWS instance metadata service (IMDS).
+
+(The name "shibaken" comes from the fact that Shiba are small, but agile, hunting dogs.)
+*/
+
+#![deny(rust_2018_idioms)]
+
+#[macro_use]
+extern crate log;
+
+use reqwest::blocking::Client;
+use serde::Serialize;
+use simplelog::{Config as LogConfig, LevelFilter, TermLogger, TerminalMode};
+use snafu::{ensure, ResultExt};
+use std::collections::HashMap;
+use std::{env, process};
+use url::Url;
+
+// Setting generators do not require dynamic socket paths at this moment.
+const DEFAULT_API_SOCKET: &str = "/run/api.sock";
+const API_METADATA_URI_BASE: &str = "/metadata/";
+
+// Instance Meta Data Service.
+const IMDS_PUBLIC_KEY_BASE_URL: &str = "http://169.254.169.254/latest/meta-data/public-keys/";
+const IMDS_SESSION_TOKEN_ENDPOINT: &str = "http://169.254.169.254/latest/api/token";
+
+#[derive(Serialize)]
+struct UserData {
+    ssh: Ssh,
+}
+
+#[derive(Serialize)]
+struct Ssh {
+    authorized_keys: Vec<String>,
+}
+
+/// Returns the value of a metadata key for a given data key, erroring if the value is not a
+/// string or is empty.
+async fn get_metadata(key: &str, meta: &str) -> Result<String> {
+    let uri = &format!("{}{}?keys={}", API_METADATA_URI_BASE, meta, key);
```

Comment:
I believe we should be querying a setting here - not metadata.  `user-data` is a setting.